### PR TITLE
Fix Observable issue where effects stop running, creating need for reloading page

### PR DESCRIPTION
**Pull request by cshashwat**
## Description

Currently, if an action such as logging in or importing a snippet fails, the Observable associated with that action is automatically unsubscribed from and its effects and reducers stop running. Future dispatches no longer work until the user manually reloads the page or clears cache.

This fix addresses the issue by using the RXJS `switchMap` operator, which creates a new Observable that is then subscribed to. This does not affect the original Observable, preserving the ability to repeatedly dispatch actions and retain functionality.

This pull request addresses known areas in the code where this is an issue:
- importing a snippet
- logging in to GitHub: https://github.com/OfficeDev/script-lab/issues/470
- sharing a gist
- adding intellisense, addressed previously in this pull request: https://github.com/OfficeDev/script-lab/pull/303/files

## Related Issue

GitHub login issue: https://github.com/OfficeDev/script-lab/issues/470

## How Has This Been Tested?

Tested by repro'ing the error, adding changes, and checking that actions are still hit and functionality is retained.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/client/app/effects/snippet.ts [hunk 1]
```diff
@@ -27,18 +27,22 @@ export class SnippetEffects {
...
-                message + '\n\n' + Strings().reloadPrompt,
-                Strings().snippetImportErrorTitle,
-                Strings().okButtonLabel)
-                .then(() => window.location.reload());
-            return Observable.from([]);
+        .switchMap(action => {
+            let data = action.payload;
+            let mode = action.mode;
+
+            return this._importRawFromSource(data, mode)
```
Comment (dsaunders45459) [inline src/client/app/effects/snippet.ts]: Can you write this as:
.ofType(Snippet.SnippetActionTypes.IMPORT) .ofType(Snippet.SnippetActionTypes.IMPORT)
.map((action: Snippet.ImportAction) => ({ data: action.payload, mode: action.mode }))
.switchMap(({ snippet, mode } => {
return this._importRawFromSource(data, mode)
})

Comment (cshashwat) [inline src/client/app/effects/snippet.ts]: Added.

**Comment by cshashwat** [comment]
@WrathOfZombies here in your pull request: https://github.com/OfficeDev/script-lab/pull/303/files, under `effects/monaco.ts`, what exactly was the error Observable that caused the chain to break? I've added a switchMap to my code in that same area, but playing with it in insiders, there doesn't seem to be any errors when loading the intellisense libraries. A user can enter incorrect libraries and then re-enter the correct ones, and the actions and reducers still get hit. What was the pull request trying to solve?

**Comment by dsaunders45459** [comment]
I looked up the difference between switchMap and mergeMap, this page does a good job of explaining it: http://javascript.tutorialhorizon.com/2017/03/29/switchmap-vs-flatmap-rxjs/. I believe we should only use switchMap instead of mergeMap when we want subscribers to ignore previous input streams when a newer event gets emitted. For login this may make sense because the user can only be signed-in with a single account, so we would take the later. For actions such as import, I think we would want to import all gists if the user starts importing a second gist before the first gist responds (you can test this scenario using fiddler autoresponder and setting 10+ second response time). Your fixing the reload issue by adding the nested catch block to not invalidate the effect chain, but I think the switchMap may be unnecessary in some of these cases. Can you re-evaluate the usage of switchMap given this definition, and make sure we are using it in the correct spots and use mergeMap where it is more suitable.

**Pull request merged**